### DISCUSSION TITLE
NER vs dep parser vs phrase matching

### DISCUSSION THREAD

**Discussion by vahuja4** [explosion/spaCy#8691] [Help: Other Questions] [answered]
Hello,

I have a corpus which consists of sentences describing apparel. Here are a couple of examples:
**This sweeping floor length beauty features a kimono tie wrap front and _free flow sleeves,_ and is finished with flat bind piping and crystal embellishments**
**Also featuring a ruched waist with silver detail finished with _sheer 1/2 length sleeves_**

From the above two sentences, as far as sleeves are concerned, I want to capture the italicized parts (everything related to sleeves). Should I use NER, dependency parsing or phrase matching? I tried dependency parsing on the first sentence and didn't work well. It did not capture the word 'free' as a dependent of 'sleeves'. I would like to understand how to decide on a technique to do this, please.

**Comment by polm** [comment, marked as answer]
Is this related to https://github.com/explosion/spaCy/discussions/8645, or is the corpus you're working with public? 

Have you tried using noun chunks? spaCy has a built in noun chunks feature that can capture phrases of the pattern `ADJ* NOUN+`, which will capture your italic phrases. It will also capture other phrases ("floor length beauty", "kimono tie wrap front", "ruched waist"), but you can use post-processing to filter them out (for example, to only keep phrases including "sleeves"). 

Dependency parsing can be useful for adjective-noun constructions, but given their structure isn't very complicated in English, matching flat tag sequences (like the Matcher can do) is also effective.

Phrase matching (as opposed to the plain Matcher) is restricted to matching things of the same token length, so it's not what you want here.

**Reply by vahuja4** [reply]
Thank you polm, this is helpful. I tried the noun chunks and it seems to work well. I will have to test more extensively though. Is there a solid reason why NER would not work for such a problem?

**Reply by vahuja4** [reply]
When I tried the noun-chunking on the sentence "This sweeping floor length beauty features a kimono tie wrap front and free flow sleeves, and is finished with flat bind piping and crystal embellishments". It is capturing this **"a kimono tie wrap front and free flow sleeves**" as a single chunk. I wonder why.

**Reply by vahuja4** [reply]
It seems to be failing whenever there is a conjunction between two phrases:
For example, "a simple scoop neckline and long balloon sleeves sleeves" is being recognized as a single chunk.

**Reply by polm** [reply]
Ah, if the noun chunks are combining things with conjunctions then that's not a bug, that's just how it works and I had forgotten. If you want you can look at the source for it and modify it to not use coordinated chunks.

For NER, you could probably train a model to recognize this, but since based on your example it looks like you're looking for a very general construction (`ADJ* NOUN+`), using matchers would be easier because then you don't need to train a model. 

There are different situations when NER is useful, but the typical one is when you have something where you could say what type it is if replaced with XXXX in a sentence, but you can't list all the possible values in advance. Personal names, locations, organizations all fall in this category. ("XXXX picked up their hat and got on the train to XXXX.") What you're looking for seems more general than that, but if I'm misunderstanding maybe you can give some more examples.

**Reply by vahuja4** [reply]
Now, I understand why NER won't work - thank you! With regards to the source code modification, could you perhaps tell me the file(s), please?

**Reply by polm** [reply]
Here's the source:

https://github.com/explosion/spaCy/blob/master/spacy/lang/en/syntax_iterators.py

Note that rather than modifying this in-place, I would copy this function and modify it as needed. You can't use the `doc.noun_chunks` if you do that, but you can just use it as a function that returns spans.

**Reply by vahuja4** [reply]
Thanks will dig into it! Meanwhile, can you please suggest a good resource to understand computational linguistics for someone who doesn't want to go into the heavy details, but enough for NLP.

**Reply by polm** [reply]
The [Jurafsky and Martin book](https://web.stanford.edu/~jurafsky/slp3/) is the best general resource, and it's written so you can skip around chapters quite freely. I still refer to it and find surprises sometimes.

**Reply by vahuja4** [reply]
Thank you @polm for your replies! Much appreciated!

**Reply by vahuja4** [reply]
@polm - can you please take a look at the sentence below and the corresponding noun-chunks. I am not able to make sense of it:

**sentence:**
 with on-trend puff sleeves and a flattering square neckline 
 **chunks:**
 `[{'chunk': 'trend', 'root': 'trend'}, {'chunk': 'a flattering square neckline', 'root': 'neckline'}]`

Would you know why `sleeves` is not being considered a noun? Can you please shed some light?

**Reply by vahuja4** [reply]
Another example:
Designed in a relaxed fabric, it comes in a relaxed shape that flows with the body.

```
CHUNK': [{'chunk': 'a relaxed fabric', 'root': 'fabric'},
   {'chunk': 'it', 'root': 'it'},
   {'chunk': 'a relaxed shape', 'root': 'shape'},
   {'chunk': 'the body', 'root': 'body'}]

```
Here, I don't understand why `it` is being considered as the root of a chunk because `it` is a pronoun, not a noun.

**Reply by polm** [reply]
Good questions here, but first as a note - the models aren't perfect, they will make mistakes, and sometimes the output will just be weird. So you should have a strategy for failing gracefully if the noun chunks are weird. 

Another question - what model are you using? I'm using the small model for testing purposes.

On the specific issues...

> with on-trend puff sleeves and a flattering square neckline

So one thing that's going on here is that there are different ways to think about "on-trend". Is the whole thing an adjective, or should it be considered a prepositional phrase that's embedded? spaCy's training data takes the latter view, so "trend" is a noun. The part of speech detection here is fine - "puff sleeves" are correctly detected as a compound. 

What's happening here is if you look at the [source](https://github.com/explosion/spaCy/blob/9b63cbb775caca77aeb6bc445a49fe0ab16d6baf/spacy/lang/en/syntax_iterators.py#L34) you can see that there's some code to avoid generating nested chunks. However, a side effect of this is that if "trend" is detected as a chunk, since it's inside the tree of "sleeves", "sleeves" can't be a chunk head. 

It might be possible to improve the way this works in noun chunks. However, if constructions like this are really important to you, you should also be able to write a Matcher rule to find them based on part of speech patterns, without reference to the dependency parse.

> Here, I don't understand why it is being considered as the root of a chunk because it is a pronoun, not a noun.

The noun chunks code considers nouns, proper nouns, and pronouns to be valid noun chunk heads. You could consider that wrong but it seems reasonable to me. If you don't want pronouns it should be easy to discard with post-processing.
